# Prosper Loan Data Exploration
## by Hao Xu

## Preliminary Wrangling


This document explores **Prosper Loan Dataset**. Prosper is America's first peer-to-peer lending company. Before exploration, I think there are two important things need to pay much attation:
1. Prosper stoped its business in 2008, and reopened their business after having obtained SEC registration for its loans on July 13 2009.
2. After Prosper's reopening, lenders simply choose whether or not to invest at the rate which Prosper's loan pricing algorithm assigns to the loan after it analyzes the borrower's credit report and financial information(no longer determine the loan rate via price discovery in an auction).

(Reference Wikipedia)

[参考](https://rpubs.com/bboylowye/282962)

In [60]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

In [61]:
# load in the dataset into a pandas dataframe,
df = pd.read_csv('prosperLoanData.csv')

In [62]:
# high-level overview of data shape and composition
print(df.shape)
print(df.dtypes)
print(df.head())

(113937, 81)
ListingKey                              object
ListingNumber                            int64
ListingCreationDate                     object
CreditGrade                             object
Term                                     int64
LoanStatus                              object
ClosedDate                              object
BorrowerAPR                            float64
BorrowerRate                           float64
LenderYield                            float64
EstimatedEffectiveYield                float64
EstimatedLoss                          float64
EstimatedReturn                        float64
ProsperRating (numeric)                float64
ProsperRating (Alpha)                   object
ProsperScore                           float64
ListingCategory (numeric)                int64
BorrowerState                           object
Occupation                              object
EmploymentStatus                        object
EmploymentStatusDuration               float64


In [63]:
# descriptive statistics for numeric variables
print(df.describe())

       ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.

In [64]:
df.ProsperScore.isnull().sum()

29084

In [65]:
df.ListingCreationDate = pd.to_datetime(df.ListingCreationDate)
df.ListingCreationDate.min()

Timestamp('2005-11-09 20:44:28.847000')

In [66]:
a = pd.Timestamp(2009,11,26)
b = pd.Timestamp(2009,7,13)
df.query('ListingCreationDate >= @b')



ListingKey  ListingNumber     ListingCreationDate  \
1       10273602499503308B223C1        1209647 2014-02-27 08:28:07.900   
3       0EF5356002482715299901A         658116 2012-10-22 11:02:35.010   
4       0F023589499656230C5E3E2         909464 2013-09-14 18:38:39.097   
5       0F05359734824199381F61D        1074836 2013-12-14 08:26:37.093   
6       0F0A3576754255009D63151         750899 2013-04-12 09:52:56.147   
...                         ...            ...                     ...   
113932  E6D9357655724827169606C         753087 2013-04-14 05:55:02.663   
113933  E6DB353036033497292EE43         537216 2011-11-03 20:42:55.333   
113934  E6E13596170052029692BB1        1069178 2013-12-13 05:49:12.703   
113935  E6EB3531504622671970D9E         539056 2011-11-14 13:18:26.597   
113936  E6ED3600409833199F711B7        1140093 2014-01-15 09:27:37.657   

       CreditGrade  Term              LoanStatus           ClosedDate  \
1              NaN    36                 Current                  NaN   
3              NaN    36                 Current                  NaN   
4              NaN    36                 Current                  NaN   
5              NaN    60                 Current                  NaN   
6              NaN    36                 Current                  NaN   
...            ...   ...                     ...                  ...   
113932         NaN    36                 Current                  NaN   
113933         NaN    36  FinalPaymentInProgress                  NaN   
113934         NaN    60                 Current                  NaN   
113935         NaN    60               Completed  2013-08-13 00:00:00   
113936         NaN    36                 Current                  NaN   

        BorrowerAPR  BorrowerRate  LenderYield  EstimatedEffectiveYield  \
1           0.12016        0.0920       0.0820                  0.07960   
3           0.12528        0.0974       0.0874                  0.08490   
4           0.24614        0.2085       0.1985                  0.18316   
5           0.15425        0.1314       0.1214                  0.11567   
6           0.31032        0.2712       0.2612                  0.23820   
...             ...           ...          ...                      ...   
113932      0.22354        0.1864       0.1764                  0.16490   
113933      0.13220        0.1110       0.1010                  0.10070   
113934      0.23984        0.2150       0.2050                  0.18828   
113935      0.28408        0.2605       0.2505                  0.24450   
113936      0.13189        0.1039       0.0939                  0.09071   

        EstimatedLoss  EstimatedReturn  ProsperRating (numeric)  \
1              0.0249          0.05470                      6.0   
3              0.0249          0.06000                      6.0   
4              0.0925          0.09066                      3.0   
5              0.0449          0.07077                      5.0   
6              0.1275          0.11070                      2.0   
...               ...              ...                      ...   
113932         0.0699          0.09500                      4.0   
113933         0.0200          0.08070                      6.0   
113934         0.1025          0.08578                      3.0   
113935         0.0850          0.15950                      4.0   
113936         0.0299          0.06081                      6.0   

       ProsperRating (Alpha)  ProsperScore  ListingCategory (numeric)  \
1                          A           7.0                          2   
3                          A           9.0                         16   
4                          D           4.0                          2   
5                          B          10.0                          1   
6                          E           2.0                          1   
...                      ...           ...                        ...   
113932                     C          

In [41]:

df2 = df[df['ListingCreationDate']>= pd.Timestamp(2009,7,13)]

In [43]:
df2.isnull().sum()

ListingKey                                 0
ListingNumber                              0
ListingCreationDate                        0
CreditGrade                            84868
Term                                       0
LoanStatus                                 0
ClosedDate                             58848
BorrowerAPR                                0
BorrowerRate                               0
LenderYield                                0
EstimatedEffectiveYield                   28
EstimatedLoss                             28
EstimatedReturn                           28
ProsperRating (numeric)                   28
ProsperRating (Alpha)                     28
ProsperScore                              28
ListingCategory (numeric)                  0
BorrowerState                              0
Occupation                              1333
EmploymentStatus                           0
EmploymentStatusDuration                  19
IsBorrowerHomeowner                        0
CurrentlyI

In [14]:
(1+0.092/24)**24-1

0.09617200601272713

In [5]:
df.columns

Index(['ListingKey', 'ListingNumber', 'ListingCreationDate', 'CreditGrade',
       'Term', 'LoanStatus', 'ClosedDate', 'BorrowerAPR', 'BorrowerRate',
       'LenderYield', 'EstimatedEffectiveYield', 'EstimatedLoss',
       'EstimatedReturn', 'ProsperRating (numeric)', 'ProsperRating (Alpha)',
       'ProsperScore', 'ListingCategory (numeric)', 'BorrowerState',
       'Occupation', 'EmploymentStatus', 'EmploymentStatusDuration',
       'IsBorrowerHomeowner', 'CurrentlyInGroup', 'GroupKey',
       'DateCreditPulled', 'CreditScoreRangeLower', 'CreditScoreRangeUpper',
       'FirstRecordedCreditLine', 'CurrentCreditLines', 'OpenCreditLines',
       'TotalCreditLinespast7years', 'OpenRevolvingAccounts',
       'OpenRevolvingMonthlyPayment', 'InquiriesLast6Months', 'TotalInquiries',
       'CurrentDelinquencies', 'AmountDelinquent', 'DelinquenciesLast7Years',
       'PublicRecordsLast10Years', 'PublicRecordsLast12Months',
       'RevolvingCreditBalance', 'BankcardUtilization',
       'Availa

### What is the structure of your dataset?

There are 113,937 records in the dataset with 81 features. Most variables are numeric in nature, but the variables cut, color, and clarity are ordered factor variables with the following levels.
(worst) ——> (best) 
cut: Fair, Good, Very Good, Premium, Ideal 
color: J, I, H, G, F, E, D 
clarity: I1, SI2, SI1, VS2, VS1, VVS2, VVS1, IF

### What is/are the main feature(s) of interest in your dataset?

I am most interested in after Prosper using a new loan pricing algorithm, the borrower rate is affected by what features.

(Many borrowers who received loans prior to 2009 (which were priced by investors) would no longer qualify for a loan, at any rate, under Prosper's new underwriting policies)

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

This data set has so much information, I think all 81 features can be devided into 3 baskets:
- Basic features of the loan: time, term, amount, rate, estimate return, category and so on;
- Pre-loan management features: borrower credit, income, loan history and so on;
- Post-loan management features: principle outstanding/loss, default, fee and so on.

I want to see how Prosper price its loan, I only want the features from the first two baskets.

Base on the relationshio between risk and gain, I expect that amount&term&use of the loan  and borrower's ability to repay will have the strongest effect on the borrower rate. The () the term and the () the amount, the lower the interest rate. Besides, the strong the ability of the borrower to repay(), the lower the interest rate. 

Besides, the borrower's willingness to repay(), the borrower's history of repayment() also affect the rate.


## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!